pkg/observer/controller.go

```diff
@@ -273,31 +273,74 @@ func (c *controller) dropInProgressNodeGroups(nodeGroups v1.NodeGroupList, cnrs
 
 // createCNRs generates and applies CNRs from the changedNodeGroups
 func (c *controller) createCNRs(changedNodeGroups []*ListedNodeGroups) {
-	klog.V(3).Infoln("applying")
-	for _, nodeGroup := range changedNodeGroups {
-		nodeNames := make([]string, 0, len(nodeGroup.List))
-		for _, node := range nodeGroup.List {
-			nodeNames = append(nodeNames, node.Name)
-		}
-		// generate cnr with prefix and use generate name method
-		cnr := generation.GenerateCNR(*nodeGroup.NodeGroup, nodeNames, c.CNRPrefix, c.Namespace)
-		generation.UseGenerateNameCNR(&cnr)
-		generation.GiveReason(&cnr, nodeGroup.Reason)
-		generation.SetAPIVersion(&cnr, apiVersion)
+    klog.V(3).Infoln("applying")
+    for _, nodeGroup := range changedNodeGroups {
+        nodeNames := make([]string, 0, len(nodeGroup.List))
+        for _, node := range nodeGroup.List {
+            nodeNames = append(nodeNames, node.Name)
+        }
+        // generate cnr with prefix and use generate name method
+        cnr := generation.GenerateCNR(*nodeGroup.NodeGroup, nodeNames, c.CNRPrefix, c.Namespace)
+        generation.UseGenerateNameCNR(&cnr)
+        generation.GiveReason(&cnr, nodeGroup.Reason)
+        generation.SetAPIVersion(&cnr, apiVersion)
+
+        name := generation.GetName(cnr.ObjectMeta)
+
+        if err := generation.ApplyCNR(c.client, c.DryMode, cnr); err != nil {
+            klog.Errorf("failed to apply cnr %q for nodegroup %q: %s", name, nodeGroup.NodeGroup.Name, err)
+        } else {
+            var drymodeStr string
+            if c.DryMode {
+                drymodeStr = "[drymode] "
+            }
+            klog.V(2).Infof("%ssuccessfully applied cnr %q for nodegroup %q", drymodeStr, name, nodeGroup.NodeGroup.Name)
+            c.CNRsCreated.WithLabelValues(nodeGroup.NodeGroup.Name).Inc()
+        }
+    }
+}
 
-		name := generation.GetName(cnr.ObjectMeta)
+// selectLowestPriorityNodeGroups returns only the node groups at the lowest priority value
+func (c *controller) selectLowestPriorityNodeGroups(changedNodeGroups []*ListedNodeGroups) []*ListedNodeGroups {
+    if len(changedNodeGroups) == 0 {
```

Comment:
this is a redundant check because `changedNodeGroups` will never be `0` here since you already check there are some prior to running this function, and if `nil` were ever to be returned here anyway that would currently crash observer because we check the first element of a `nil` list right after the call of this function.

Reply:
Yes please create and return an empty list if you end up keeping this check.